Wpe 2.38 quirks

philn opened the pull request.

philn committed:
[GStreamer][WPE] Replace platform ifdefs with runtime quirks
https://bugs.webkit.org/show_bug.cgi?id=269287

Reviewed by NOBODY (OOPS!).

Compile-time quirks prevent re-usability of the same binaries across different platforms (using
containers). The proposed solution is to always compile those, they don't depend on any specific
APIs so that shouldn't be an issue. Then at runtime quirks can be tweaked using the
WEBKIT_GST_QUIRKS and WEBKIT_GST_HOLE_PUNCH environment variables.

* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
(WebCore::AudioDestinationGStreamer::AudioDestinationGStreamer):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::decodebinAutoplugSelectCallback):
(WebCore::AudioFileReader::~AudioFileReader):
(WebCore::AudioFileReader::decodeAudioForBusCreation):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::ensureGStreamerInitialized):
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::ElementFactories::ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForCaps const):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
(WebCore::MediaPlayerPrivateGStreamer::~MediaPlayerPrivateGStreamer):
(WebCore::setSyncOnClock):
(WebCore::MediaPlayerPrivateGStreamer::createAudioSink):
(WebCore::MediaPlayerPrivateGStreamer::gstreamerPositionFromSinks const):
(WebCore::MediaPlayerPrivateGStreamer::naturalSize const):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::configureElement):
(WebCore::MediaPlayerPrivateGStreamer::configureElementPlatformQuirks):
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin):
(WebCore::MediaPlayerPrivateGStreamer::swapBuffersIfNeeded):
(WebCore::setRectangleToVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::isHolePunchRenderingEnabled const):
(WebCore::MediaPlayerPrivateGStreamer::createHolePunchVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::shouldIgnoreIntrinsicSize):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkBcmNexus.cpp: Added.
(WebCore::GStreamerHolePunchQuirkBcmNexus::setHolePunchVideoRectangle):
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkBcmNexus.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkWesteros.cpp: Added.
(WebCore::GStreamerHolePunchQuirkWesteros::createHolePunchVideoSink):
(WebCore::GStreamerHolePunchQuirkWesteros::setHolePunchVideoRectangle):
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkWesteros.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp: Added.
(WebCore::GStreamerQuirkAmLogic::GStreamerQuirkAmLogic):
(WebCore::GStreamerQuirkAmLogic::createWebAudioSink):
(WebCore::GStreamerQuirkAmLogic::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBcmNexus.cpp: Added.
(WebCore::GStreamerQuirkBcmNexus::GStreamerQuirkBcmNexus):
(WebCore::GStreamerQuirkBcmNexus::isHardwareAccelerated):
* Source/WebCore/platform/gstreamer/GStreamerQuirkBcmNexus.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp: Added.
(WebCore::GStreamerQuirkBroadcom::GStreamerQuirkBroadcom):
(WebCore::GStreamerQuirkBroadcom::configureElement):
(WebCore::GStreamerQuirkBroadcom::isHardwareAccelerated):
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.cpp: Added.
(WebCore::GStreamerQuirkRealtek::GStreamerQuirkRealtek):
(WebCore::GStreamerQuirkRealtek::createWebAudioSink):
(WebCore::GStreamerQuirkRealtek::configureElement):
(WebCore::GStreamerQuirkRealtek::isHardwareAccelerated):
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp: Added.
(WebCore::GStreamerQuirkWesteros::GStreamerQuirkWesteros):
(WebCore::GStreamerQuirkWesteros::configureElement):
(WebCore::GStreamerQuirkWesteros::isHardwareAccelerated):
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.h: Added.
* Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp: Added.
(WebCore::GStreamerQuirksManager::singleton):
(WebCore::GStreamerQuirksManager::GStreamerQuirksManager):
(WebCore::GStreamerQuirksManager::isEnabled const):
(WebCore::GStreamerQuirksManager::createWebAudioSink):
(WebCore::GStreamerQuirksManager::createHolePunchVideoSink):
(WebCore::GStreamerQuirksManager::setHolePunchVideoRectangle):
(WebCore::GStreamerQuirksManager::configureElement):
(WebCore::GStreamerQuirksManager::isHardwareAccelerated const):
(WebCore::GStreamerQuirksManager::supportsVideoHolePunchRendering const):
(WebCore::GStreamerQuirksManager::audioVideoDecoderFactoryListType const):
(WebCore::GStreamerQuirksManager::disallowedWebAudioDecoders const):
* Source/WebCore/platform/gstreamer/GStreamerQuirks.h: Added.
(WebCore::GStreamerQuirk::isPlatformSupported const):
(WebCore::GStreamerQuirk::createWebAudioSink):
(WebCore::GStreamerQuirk::configureElement):
(WebCore::GStreamerQuirk::isHardwareAccelerated):
(WebCore::GStreamerQuirk::audioVideoDecoderFactoryListType const):
(WebCore::GStreamerQuirk::disallowedWebAudioDecoders const):
(WebCore::GStreamerHolePunchQuirk::createHolePunchVideoSink):
(WebCore::GStreamerHolePunchQuirk::setHolePunchVideoRectangle):
* Source/cmake/OptionsWPE.cmake:

philn committed:
Quirks: Fix build after backporting

philn committed:
Quirks: Allow legacy platform ifdefs

philn committed:
Quirks: Fix hole-punch rendering

philn committed:
Quirks: Fix non-hole punch rendering

philn committed:
Quirks: Add fakevideosink hole-punch quirk

philn committed:
Quirks: set video rectangle even if no platform quirk is requested

philn committed:
Quirks: Fix debug build

philn committed:
Quirks: Fix fake hole-punching

philn committed:
Quirks: Fix build

philn committed:
Quirks: Add plumbing for hole-punch testing

philn commented:
Hold on, I forgot to remove one ifdef...

philn committed:
Quirks: Add quirk for sink clock sync'ing

philn committed:
Quirks: Review feedback changes

philn commented:
> Hold on, I forgot to remove one ifdef...

This fixed, the PR is ready for testing/review!

emutavchi reviewed (commented):
Quirks should also allow changing playbin flags in `MediaPlayerPrivateGStreamer::setPlaybackFlags` . For example native-audio is Broadcom quirk.

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -0,0 +1,112 @@
+
+    virtual bool isPlatformSupported() const { return true; }
+    virtual GstElement* createWebAudioSink() { return nullptr; }
+    virtual bool configureElement(GstElement*, const OptionSet<ElementRuntimeCharacteristics>&) { return false; }
```
Would you mind exposing pipeline configuration via Quirk as well? Currently we maintain a patch to set custom context on the pipeline to signal limited resource usage configuration: https://code.rdkcentral.com/r/plugins/gitiles/rdk/components/generic/rdk-oe/meta-rdk-ext/+/refs/heads/rdk-next/recipes-extended/wpe-webkit/files/2.38/comcast-XIONE-12272-configure-video-resource-usage-w.patch#22, it'd be great to have an option to put it into Westeros quirk.

philn replied:
Yeah I can add this code in the Westeros quirk, no problem I think.

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp:
```diff
@@ -0,0 +1,87 @@
+    if (g_str_has_prefix(GST_ELEMENT_NAME(element), "uridecodebin3")) {
+        GRefPtr<GstCaps> defaultCaps;
+        g_object_get(element, "caps", &defaultCaps.outPtr(), nullptr);
+        defaultCaps = adoptGRef(gst_caps_merge(gst_caps_ref(m_sinkCaps.get()), defaultCaps.leakRef()));
+        GST_INFO("Setting stop caps to %" GST_PTR_FORMAT, defaultCaps.get());
+        g_object_set(element, "caps", defaultCaps.get(), nullptr);
```
I believe this was removed in favor for change in playbin3 for supporting decoding sinks. For reference: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1093. I guess it should be safe to keep this in Westeros quirk until playbin3 changes are sorted out.

philn replied:
Hmm the current code comes from upstream's ifdefs... Was there some mixup in the upstreaming of #1093 @calvaris ?

philn replied:
Hopefully all the comments excepted this one should be addressed in https://github.com/WebKit/WebKit/pull/26183

calvaris replied:
It was not removed, it won't change upstream and after discussing it with GStreamer upstream developers it is the way to go, so current code is correct.

calvaris replied:
Ah, this is for 2.38, I was thinking on other versions. For 2.38, the GStreamer patch drafted in https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4391 is in, so we don't need to tweak the template caps. This would be the last version not needing to tweak the caps, from 2.42 onwards, we need to tweak the template caps as the GStreamer patch won't be present.

calvaris replied:
Summing up, I think @emutavchi is right and we need to remove this from here.

philn replied:
Done.

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp:
```diff
@@ -0,0 +1,65 @@
+bool GStreamerQuirkAmLogic::configureElement(GstElement* element, const OptionSet<ElementRuntimeCharacteristics>& characteristics)
+{
+    GST_INFO("Set property disable-xrun to TRUE");
+    g_object_set(element, "disable-xrun", TRUE, nullptr);
```
This should check and configure only GstAmlHalAsink. Otherwise it produces a lot of noise:

```
Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.940: g_object_set_is_valid_property: object class 'GstURIDecodeBin' has no property named 'disable-xrun'                                                                                                                   Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.942: g_object_set_is_valid_property: object class 'WebKitWebSrc' has no property named 'disable-xrun'                                                                                                                      
Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.942: g_object_set_is_valid_property: object class 'GstTypeFindElement' has no property named 'disable-xrun'                                                                                
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstDecodeBin' has no property named 'disable-xrun'
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstTypeFindElement' has no property named 'disable-xrun'                                                                                                                
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstQueue2' has no property named 'disable-xrun'                                                                                                                         
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.334: g_object_set_is_valid_property: object class 'GstQTDemux' has no property named 'disable-xrun'                                                                                                                        
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.337: g_object_set_is_valid_property: object class 'GstMultiQueue' has no property named 'disable-xrun'                                                                                                                     
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.339: g_object_set_is_valid_property: object class 'GstH264Parse' has no property named 'disable-xrun'                                                                                                                      
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.341: g_object_set_is_valid_property: object class 'GstCapsFilter' has no property named 'disable-xrun'                                                                                                                     
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.344: g_object_set_is_valid_property: object class 'avdec_h264' has no property named 'disable-xrun'                                                                                                                        
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.349: g_object_set_is_valid_property: object class 'GstAacParse' has no property named 'disable-xrun'                                                                                                                       
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.374: g_object_set_is_valid_property: object class 'avdec_aac' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.401: g_object_set_is_valid_property: object class 'GstInputSelector' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.402: g_object_set_is_valid_property: object class 'GstInputSelector' has no property named 'disable-xrun'                                                                                                                  Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.403: g_object_set_is_valid_property: object class 'GstTee' has no property named 'disable-xrun'                                                                                                                            
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'GstBin' has no property named 'disable-xrun'   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'GstQueue' has no property named 'disable-xrun'                                                                                                                          
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'WebKitVideoSink' has no property named 'disable-xrun'                                                                                                                   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.408: g_object_set_is_valid_property: object class 'GstBin' has no property named 'disable-xrun'                                                                                                                            
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.408: g_object_set_is_valid_property: object class 'GstPlaySinkAudioConvert' has no property named 'disable-xrun'                                                                                                           
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstAudioResample' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstAudioConvert' has no property named 'disable-xrun'                                                                                                                   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstIdentity' has no property named 'disable-xrun'                                                                                                                       
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstQueue' has no property named 'disable-xrun' 
```

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp:
```diff
@@ -0,0 +1,65 @@
+    if (characteristics.contains(ElementRuntimeCharacteristics::HasVideo))
+        g_object_set(element, "wait-video", TRUE, nullptr);
+
+    return true;
```
I think it should return false for non GstAmlHalAsink  for other Quirks to kick in

emutavchi commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4180,11 +4143,13 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
+        if (m_videoSink) {
+            pushNextHolePunchBuffer();
+            return m_videoSink.get();
+        }
```
It should return null here even if hole punching quirk did not return anything. This will preserve existing behavior we have, where we rely on playbin2 to find a suitable sink for us.

philn commented:
> Quirks should also allow changing playbin flags in `MediaPlayerPrivateGStreamer::setPlaybackFlags` . For example native-audio is Broadcom quirk.

I wasn't aware of that. Yeah definitely, @calvaris already suggested removing those native-video/etc ifdefs. I can take a look.

philn committed:
Quirks: Address review feedback

Backport of https://commits.webkit.org/276450@main keeping old ifdefs
compatibility, so the GStreamerQuirksManager::getAdditionalPlaybinFlags()
implementation diverges from upstream version.

philn commented:
I've rebased the branch and backported the changes requested by @emutavchi (see most recent commit).

philn committed:
Quirks: Westeros: Remove uridecodebin3 configuration

Downstream GStreamer is already patched for playbin3 decoder/sink handling.

calvaris reviewed (approved):
LGTM. What I don't know is if we should wait until we have a confirmation from @emutavchi .

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp:
```diff
@@ -0,0 +1,73 @@
+        g_object_set(element, "low_latency", TRUE, "low_latency_max_queued_ms", 60, nullptr);
+    }
+
+    return true;
```
This is similar to Amlogic quirk complain in previous review.  Returning 'true' here breaks the loop in GStreamerQuirksManager::configureElement(), blocking other potential quirks. On RDK we'll usually have 2 quirks active: platform specific (amlogic or realtek or broadcom) and westeros. Soc specific quirk should not block Westeros or other potential quirks If an element is not recognized by a quirk.  Also, it could be useful for multiple quirks to be able to configure same element (we don't have such use case today though).

philn replied:
OK, makes sense I think :) Have you tested the suggested changes on the various platforms?

philn replied:
Maybe the `GStreamerQuirksManager::configureElement()` could be re-worked to apply every quirk instead of early loop breaks...

philn replied:
Does this make sense? https://github.com/WebKit/WebKit/pull/26456

emutavchi replied:
> OK, makes sense I think :) Have you tested the suggested changes on the various platforms?

I did test on Amlogic and Realtek so far. Results looks good. I plan to test Broadcom later this week. 

> Does this make sense? https://github.com/WebKit/WebKit/pull/26456

yes, that works for me.

emutavchi replied:
@philn, I just noticed that quirks doesn't cover libwebrtc GStreamerVideoDecoderFactory.cpp. Specifically  the code that skips hw elements in passthrough pipeline: https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoDecoderFactory.cpp#L72

and h264 parser requirement https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoDecoderFactory.cpp#L355

Would you mind adding it to this PR?

philn replied:
Hmm I wasn't aware of those, seems like they were not upstreamed. Thanks for the heads-up, I'll check this tomorrow.

philn replied:
Can you test this? https://github.com/WebKit/WebKit/pull/26506

emutavchi replied:
> Can you test this? https://github.com/WebKit/WebKit/pull/26506

Hi @philn, I did test it on Amlogic, Realtek, Broadcom. I took https://github.com/WebKit/WebKit/pull/26456 and https://github.com/WebKit/WebKit/pull/26506 on top of current PR. Results looks good.

philn replied:
I've backported these two here.

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.cpp:
```diff
@@ -0,0 +1,88 @@
+        g_object_set(element, "lowdelay-mode", TRUE, nullptr);
+    }
+
+    return true;
```
Same here.

emutavchi commented on Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp:
```diff
@@ -0,0 +1,78 @@
+        GST_INFO("Enable 'immediate-output' in WesterosSink");
+        g_object_set(element, "immediate-output", TRUE, nullptr);
+    }
+    return true;
```
And here.

emutavchi reviewed (changes requested):
Few more changes required, that were overlooked last time.

philn committed:
[GStreamer][Quirks] Apply configureElement to every requested quirk
https://bugs.webkit.org/show_bug.cgi?id=271697

Reviewed by NOBODY (OOPS!).

On some platforms it is desirable to apply that quirk unconditionally, because multiple
platform-specific quirks are requested.

* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp:
(WebCore::GStreamerQuirkAmLogic::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp:
(WebCore::GStreamerQuirkBroadcom::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.cpp:
(WebCore::GStreamerQuirkRealtek::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp:
(WebCore::GStreamerQuirkWesteros::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp:
(WebCore::GStreamerQuirksManager::configureElement):
* Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
(WebCore::GStreamerQuirk::configureElement):

philn committed:
[GStreamer][LibWebRTC] Add quirks in decoder factory
https://bugs.webkit.org/show_bug.cgi?id=271751

Reviewed by NOBODY (OOPS!).

Allow the decoder pipeline to skip auto-plugging of hardware-accelerated decoders and also to
disable parsing, both for Broadcom and Realtek quirks.

* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.h:
* Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp:
(WebCore::GStreamerQuirksManager::shouldParseIncomingLibWebRTCBitStream const):
* Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
(WebCore::GStreamerQuirk::shouldParseIncomingLibWebRTCBitStream const):
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoDecoderFactory.cpp:
(WebCore::GStreamerWebRTCVideoDecoder::getGstAutoplugSelectResult):
(WebCore::H264Decoder::H264Decoder):

modeveci commented:
@emutavchi , Once you approved this PR, then I think we are ready to merge. @philn let's wait Eugene's approval and then we can continue.

emutavchi reviewed (approved):
Looks good to me
